samples/aws-codestar-lambda/src/pedestallambda/service.clj

```diff
@@ -2,9 +2,83 @@
   (:require [io.pedestal.http :as http]
             [io.pedestal.http.route :as route]
             [io.pedestal.http.body-params :as body-params]
-            [ring.util.response :as ring-resp]))
+            [io.pedestal.http.impl.servlet-interceptor :as servlet-utils]
+            [io.pedestal.interceptor :as interceptor]
+            [cheshire.core :as json]
+            [ring.util.response :as ring-resp])
+  (:import (java.io ByteArrayOutputStream)))
 
-;; Our service returns EDN that is sent as JSON bodies
+;; PATCHING
+;; ------------------------------------------
+;;
+;; `reduce-kv` doesn't drop back to generic `java.util.Map` support
+;; like `reduce` and other sequence-oriented functions in clojure.core.
+;; This is a known issue in Clojure: https://dev.clojure.org/jira/browse/CLJ-1762
+;;
+;; This is needed because Pedestal uses `reduce-kv` when processing Map-like
+;; requests and APIGW uses java.util.Maps.
+;;
+;; THIS SHOULD BE REMOVED ONCE CLOJURE IS PATCHED!
+;; ----
+(extend-protocol clojure.core.protocols/IKVReduce
+  java.util.Map
+  (kv-reduce
+    [amap f init]
+    (let [^java.util.Iterator iter (.. amap entrySet iterator)]
+      (loop [ret init]
+        (if (.hasNext iter)
+          (let [^java.util.Map$Entry kv (.next iter)
+                ret (f ret (.getKey kv) (.getValue kv))]
+            (if (reduced? ret)
+              @ret
+              (recur ret)))
+          ret)))))
+;; ------------------------------------------
+
+
+;; Interceptors
+;; ------------------------------------------
+(def pre-body-params
+  "An interceptor that ensures `:content-type` is on the Request map/object.
+  This is achieved by first looking to see if the key already exists, and if not,
+  adding the information as reported in the Content-Type header.
+  This is necessary because `body-params` expects `:content-type`, but some
+  chain providers (like APIGW) do the bare minimum when making the request map/object."
+  (interceptor/interceptor
+    {:name  ::pre-body-params
+     :enter (fn [ctx]
+              (if (get-in ctx [:request :content-type])
+                ctx
+                (assoc-in ctx [:request :content-type] (or (get-in ctx [:request :headers "content-type"])
```

Comment:
Based on the doc-string I gather that the request object may not be a clojure map. Will it be `associative?`? Otherwise this won't work, will it?

Reply:
Every one of Pedestal's request-map implementations supports Associative (including the one-off/standalone implementation used in the Lambda/APIGW provider).  This works as expected.  The "object/map" is a reference that it could be satisfy the Associative interface but be some other object underneath (other than a persistent map, as seen in the zerocopy request).

All of Clojure's core abstractions are interfaces (maps, seqs, associative), so `assoc` never implies a concrete data type, just something that implements the appropriate interfaces.

Reply:
👍